wagtail/contrib/search_promotions/views.py

```diff
@@ -6,80 +6,73 @@
-    is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-            },
-        )
+class SearchPromotionColumn(RelatedObjectsColumn):
+    cell_template_name = "wagtailsearchpromotions/search_promotion_column.html"
+
+
+class IndexView(generic.IndexView):
+    model = Query
+    template_name = "wagtailsearchpromotions/index.html"
+    results_template_name = "wagtailsearchpromotions/index_results.html"
+    context_object_name = "queries"
+    page_title = gettext_lazy("Search Terms")
+    header_icon = "pick"
+    paginate_by = 20
+    permission_policy = ModelPermissionPolicy(SearchPromotion)
+    index_url_name = "wagtailsearchpromotions:index"
+    index_results_url_name = "wagtailsearchpromotions:index_results"
+    _show_breadcrumbs = True
+    search_fields = ["query_string"]
+    add_url_name = "wagtailsearchpromotions:add"
+    add_item_label = gettext_lazy("Add new promoted result")
+    columns = [
+        TitleColumn(
+            "query_string",
+            label=gettext_lazy("Search term(s)"),
+            width="40%",
+            url_name="wagtailsearchpromotions:edit",
+            sort_key="query_string",
+        ),
+        SearchPromotionColumn(
+            "editors_picks",
+            label=gettext_lazy("Promoted results"),
+            width="40%",
+        ),
+        Column(
+            "views",
+            label=gettext_lazy("Views (past week)"),
+            width="20%",
+            sort_key="views",
+        ),
+    ]
+
+    def get_base_queryset(self):
+        has_promotions = SearchPromotion.objects.values_list("query_id", flat=True)
+        queryset = self.model.objects.filter(pk__in=has_promotions)
+        queryset = queryset.annotate(
+            views=functions.Coalesce(Sum("daily_hits__hits"), 0)
+        ).prefetch_related("editors_picks", "editors_picks__page")
```

Comment:
It'd also be great to have a test case for the number of queries on the index page when we have `Query` objects with multiple `SearchPromotion`s, so we can clearly see how significant the improvement is.